### PR TITLE
Clear tree caches when the compilation unit changes (Fixes #919)

### LINKED ISSUE #919: Compilation failure in 2.1.3
I have a project that compiles fine with 2.1.2, but fails with 2.1.3. When running via Maven I don't see any output from the compiler other than this:

```
Failed to execute goal org.apache.maven.plugins:maven-compiler-plugin:3.5.1:compile (default-compile) on project ...: Compilation failure
```

If I extract the actual javac command from mvn -X and run it manually I see this:

```
[checking ...my class]
[loading ZipFileIndexFileObject[/Library/Java/JavaVirtualMachines/jdk1.8.0_101.jdk/Contents/Home/lib/ct.sym(META-INF/sym/rt.jar/java/net/URI.class)]]
[loading ZipFileIndexFileObject[/...my lib]]
[loading ZipFileIndexFileObject[/.../.m2/repository/org/checkerframework/jdk8/2.1.3/jdk8-2.1.3.jar(java/util/function/Function.class)]]
[loading ZipFileIndexFileObject[/.../.m2/repository/org/checkerframework/jdk8/2.1.3/jdk8-2.1.3.jar(java/util/Comparator.class)]]
[loading ZipFileIndexFileObject[/.../.m2/repository/org/checkerframework/jdk8/2.1.3/jdk8-2.1.3.jar(java/lang/Number.class)]]
[loading ZipFileIndexFileObject[/.../.m2/repository/org/checkerframework/jdk8/2.1.3/jdk8-2.1.3.jar(java/util/AbstractSet.class)]]
[loading ZipFileIndexFileObject[/.../.m2/repository/org/checkerframework/jdk8/2.1.3/jdk8-2.1.3.jar(java/io/IOException.class)]]
[loading ZipFileIndexFileObject[/.../.m2/repository/org/checkerframework/checker-qual/2.1.3/checker-qual-2.1.3.jar(org/checkerframework/checker/nullness/qual/NonNull.class)]]
[loading ZipFileIndexFileObject[/Library/Java/JavaVirtualMachines/jdk1.8.0_101.jdk/Contents/Home/lib/ct.sym(META-INF/sym/rt.jar/javax/lang/model/type/TypeKind.class)]]
[loading ZipFileIndexFileObject[/.../.m2/repository/org/checkerframework/checker-qual/2.1.3/checker-qual-2.1.3.jar(org/checkerframework/framework/qual/DefaultInUncheckedCodeFor.class)]]
[loading ZipFileIndexFileObject[/Library/Java/JavaVirtualMachines/jdk1.8.0_101.jdk/Contents/Home/lib/ct.sym(META-INF/sym/rt.jar/java/lang/SafeVarargs.class)]]
error: ElementAnnotationUtil.getLocationTypeADT: invalid location TYPE_ARGUMENT(0) for type: Set
  Compilation unit: /...snip.java
  Exception: java.lang.Throwable; Stack trace: org.checkerframework.framework.source.SourceChecker.errorAbort(SourceChecker.java:744)
  org.checkerframework.javacutil.ErrorReporter.errorAbort(ErrorReporter.java:28)
  org.checkerframework.framework.util.element.ElementAnnotationUtil.getLocationTypeADT(ElementAnnotationUtil.java:432)
  org.checkerframework.framework.util.element.ElementAnnotationUtil.getTypeAtLocation(ElementAnnotationUtil.java:361)
  org.checkerframework.framework.util.element.ElementAnnotationUtil.annotateViaTypeAnnoPosition(ElementAnnotationUtil.java:276)
  org.checkerframework.framework.util.element.ParamApplier.handleTargeted(ParamApplier.java:221)
  org.checkerframework.framework.util.element.TargetedElementAnnotationApplier.extractAndApply(TargetedElementAnnotationApplier.java:189)
  org.checkerframework.framework.util.element.ParamApplier.extractAndApply(ParamApplier.java:259)
  org.checkerframework.framework.util.element.ParamApplier.apply(ParamApplier.java:51)
  org.checkerframework.framework.type.ElementAnnotationApplier.apply(ElementAnnotationApplier.java:96)
  org.checkerframework.framework.type.TypeFromMemberVisitor.visitVariable(TypeFromMemberVisitor.java:34)
  org.checkerframework.framework.type.TypeFromMemberVisitor.visitVariable(TypeFromMemberVisitor.java:23)
  com.sun.tools.javac.tree.JCTree$JCVariableDecl.accept(JCTree.java:864)
  com.sun.source.util.SimpleTreeVisitor.visit(SimpleTreeVisitor.java:53)
  org.checkerframework.framework.type.TypeFromTree.fromMember(TypeFromTree.java:51)
  org.checkerframework.framework.type.AnnotatedTypeFactory.fromMember(AnnotatedTypeFactory.java:1213)
  org.checkerframework.framework.type.AnnotatedTypeFactory.fromElement(AnnotatedTypeFactory.java:1127)
  org.checkerframework.framework.type.AnnotatedTypeFactory.getAnnotatedType(AnnotatedTypeFactory.java:976)
  org.checkerframework.framework.flow.CFAbstractTransfer.getValueFromFactory(CFAbstractTransfer.java:195)
  org.checkerframework.framework.flow.CFAbstractTransfer.visitLocalVariable(CFAbstractTransfer.java:645)
  org.checkerframework.framework.flow.CFAbstractTransfer.visitLocalVariable(CFAbstractTransfer.java:97)
  org.checkerframework.dataflow.cfg.node.LocalVariableNode.accept(LocalVariableNode.java:76)
  org.checkerframework.dataflow.analysis.Analysis.callTransferFunction(Analysis.java:404)
  org.checkerframework.dataflow.analysis.Analysis.performAnalysis(Analysis.java:197)
  org.checkerframework.framework.type.GenericAnnotatedTypeFactory.analyze(GenericAnnotatedTypeFactory.java:990)
  org.checkerframework.framework.type.GenericAnnotatedTypeFactory.analyze(GenericAnnotatedTypeFactory.java:947)
  org.checkerframework.framework.type.GenericAnnotatedTypeFactory.performFlowAnalysis(GenericAnnotatedTypeFactory.java:874)
  org.checkerframework.framework.type.GenericAnnotatedTypeFactory.checkAndPerformFlowAnalysis(GenericAnnotatedTypeFactory.java:1189)
  org.checkerframework.framework.type.GenericAnnotatedTypeFactory.preProcessClassTree(GenericAnnotatedTypeFactory.java:208)
  org.checkerframework.common.basetype.BaseTypeVisitor.visitClass(BaseTypeVisitor.java:286)
  org.checkerframework.common.basetype.BaseTypeVisitor.visitClass(BaseTypeVisitor.java:169)
  com.sun.tools.javac.tree.JCTree$JCClassDecl.accept(JCTree.java:720)
  com.sun.source.util.TreePathScanner.scan(TreePathScanner.java:50)
  org.checkerframework.framework.source.SourceVisitor.visit(SourceVisitor.java:66)
  org.checkerframework.framework.source.SourceChecker.typeProcess(SourceChecker.java:978)
  org.checkerframework.common.basetype.BaseTypeChecker.typeProcess(BaseTypeChecker.java:464)
  org.checkerframework.javacutil.AbstractTypeProcessor$AttributionTaskListener.finished(AbstractTypeProcessor.java:208)
  com.sun.tools.javac.api.ClientCodeWrapper$WrappedTaskListener.finished(ClientCodeWrapper.java:681)
  com.sun.tools.javac.api.MultiTaskListener.finished(MultiTaskListener.java:111)
  com.sun.tools.javac.main.JavaCompiler.flow(JavaCompiler.java:1342)
  com.sun.tools.javac.main.JavaCompiler.flow(JavaCompiler.java:1296)
  com.sun.tools.javac.main.JavaCompiler.compile2(JavaCompiler.java:901)
  com.sun.tools.javac.main.JavaCompiler.compile(JavaCompiler.java:860)
  com.sun.tools.javac.main.Main.compile(Main.java:523)
  com.sun.tools.javac.main.Main.compile(Main.java:381)
  com.sun.tools.javac.main.Main.compile(Main.java:370)
  com.sun.tools.javac.main.Main.compile(Main.java:361)
  com.sun.tools.javac.Main.compile(Main.java:56)
  com.sun.tools.javac.Main.main(Main.java:42)
```

The problem seemed intermittent, but I figured out it depends on the order the source files were passed to the compiler. Certain source file orderings cause the above error, other source orderings actually work.

### PR DESCRIPTION
**Pull request by smillst**
Fixes #919

### CODE REVIEW

File: framework/src/org/checkerframework/framework/type/AnnotatedTypeFactory.java [hunk 1]
```diff
@@ -302,7 +302,7 @@
     private final static int DEFAULT_CACHE_SIZE = 300;
 
     /** Mapping from a Tree to its annotated type; implicits have been applied. */
```
Comment (mernst) [inline framework/src/org/checkerframework/framework/type/AnnotatedTypeFactory.java]: The name suggests that the key must be a ClassTree or a MethodTree.  Is that true?  If so, I would so note in the Javadoc comment.

File: framework/src/org/checkerframework/framework/type/AnnotatedTypeFactory.java [hunk 2]
```diff
@@ -521,15 +521,15 @@ public void setRoot(/*@Nullable*/ CompilationUnitTree root) {
...
-        fromTreeCache.clear();
-        elementCache.clear();
+        // Clear the caches with trees because once the compilation unit changes,
+        // the trees may be modified and lose type arguments.
         elementToTreeCache.clear();
-        */
+        fromTreeCache.clear();
+        classAndMethodTreeCache.clear();
+
+        // There is no need to clear the following cache, it is limited by cache size and it
```
Comment (mernst) [inline framework/src/org/checkerframework/framework/type/AnnotatedTypeFactory.java]: it -> its

**Comment by mernst** [comment]
Looks good, modulo two tiny comments.

**Comment by wmdietl** [comment]
Looks good.

**Pull request merged**